api_tests/users/views/test_user_claim.py

```diff
@@ -0,0 +1,257 @@
+        # PATCH
+        res = app.patch_json_api(
+            url.format(unreg_user._id),
+            payload,
+            expect_errors=True)
+        assert res.status_code == 405
+        res = app.patch_json_api(
+            url.format(unreg_user._id),
+            payload,
+            auth=claimer.auth,
+            expect_errors=True)
+        assert res.status_code == 405
+        # PUT
+        res = app.put_json_api(
+            url.format(unreg_user._id),
+            payload,
+            expect_errors=True)
+        assert res.status_code == 405
+        res = app.put_json_api(
+            url.format(unreg_user._id),
+            payload,
+            auth=claimer.auth,
+            expect_errors=True)
+        assert res.status_code == 405
+        # DELETE
+        res = app.delete(
+            url.format(unreg_user._id),
+            expect_errors=True)
+        assert res.status_code == 405
+        res = app.delete(
+            url.format(unreg_user._id),
+            auth=claimer.auth,
+            expect_errors=True)
+        assert res.status_code == 405
+
+    def test_claim_unauth_failure(self, app, url, unreg_user, project, wrong_preprint):
+        _url = url.format(unreg_user._id)
+        # no record locator
+        payload = self.payload(email='david@david.son')
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 400
+        assert res.json['errors'][0]['detail'] == 'Must specify record "id".'
+
+        # bad record locator
+        payload = self.payload(email='david@david.son', id='notaguid')
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 404
+        assert res.json['errors'][0]['detail'] == 'Unable to find specified record.'
+
+        # wrong record locator
+        payload = self.payload(email='david@david.son', id=wrong_preprint._id)
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 404
+        assert res.json['errors'][0]['detail'] == 'Unable to find specified record.'
+
+        # no email
+        payload = self.payload(id=project._id)
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 400
+        assert res.json['errors'][0]['detail'] == 'Must either be logged in or specify claim email.'
+
+        # active user
+        _url = url.format(project.creator._id)
+        payload = self.payload(email=project.creator.email, id=project._id)
+        res = app.post_json_api(
+            _url,
+            payload,
+            expect_errors=True
+        )
+        assert res.status_code == 401
+
+    @mock.patch('website.project.views.contributor.mails.send_mail')
```

Comment:
Nit: This could be made into a fixture

```python
@pytest.fixture
def mock_mail(self):
    with mock.patch('website.project.views.contributor.mails.send_mail') as patch:
        yield patch
```

Reply:
Using this as a fixture can also capture "Welcome"-type mails sent at user creation, but using it as the last fixture avoids this problem